File: dart/structural/composite/graphic/bin/compound_graphic.dart
import './graphic.dart';

class CompoundGraphic implements Graphic {
  List<Graphic> children;

  CompoundGraphic() {
    children = [];
  }

  void add(Graphic child) {
    children.add(child);
  }

  void remove(Graphic child) {
    children.remove(child);
  }

  @override
  void draw() {
    children.forEach((v) {
      v.draw();
    });
  }

  @override
  void move(int x, int y) {
    children.forEach((v) {
      v.move(x, y);
    });
  }
}

File: typescript/structural/composite/graphic/compound_graphic.ts
import {Graphic} from "./graphic";

export class CompoundGraphic implements Graphic {
  children: Graphic[]

  constructor() {
    this.children = []
  }

  add(child: Graphic): void {
    this.children.push(child)
    this.draw()
  }

  remove(child: Graphic): void {
    const childIndex = this.children.indexOf(child)
    this.children.splice(childIndex, 1)
    this.draw()
  }

  move(x: number, y: number) {
    this.children.forEach(v => {
      v.move(x, y)
    })
    this.draw()
  }

  draw() {
    this.children.forEach(v => {
      v.draw()
    })
  }
}
